File: backend/models/account.py
from sqlalchemy import Column, Integer, String, Float, Enum, ForeignKey
from sqlalchemy.orm import relationship
from config.db.database import Base
import enum


class AccountType(enum.Enum):
    cash = "cash"
    bank = "bank"


class Account(Base):
    __tablename__ = "accounts"

    # Primary key for Account
    id = Column(Integer, primary_key=True, index=True)
    # Foreign key relationship with User (assuming there's a User table)
    user_id = Column(
        Integer, ForeignKey("users.user_id"), nullable=False
    )  # Foreign key to User

    # Other fields
    account_type = Column(Enum(AccountType), nullable=False)  # cash or bank
    bank_name = Column(String, nullable=True)  # Only applicable for bank accounts
    account_name = Column(String, nullable=False)  # e.g., 'John Doe Checking'
    account_number = Column(Integer, nullable=True)  # Bank account number
    credit = Column(Float, default=0.0)
    debit = Column(Float, default=0.0)
    balance = Column(Float, default=0.0)  # Current balance of the account

    # Relationship to User
    user = relationship("User", back_populates="accounts")

    # Relationship to Transaction (one-to-many)
    transactions = relationship(
        "Transaction", back_populates="account", cascade="all, delete-orphan"
    )

    def __repr__(self):
        return f"<Account(id={self.id}, account_name={self.account_name}, balance={self.balance})>"

File: frontend/lib/models/Account.dart
enum AccountType { bank, cash }

extension AccountTypeExtension on AccountType {
  static AccountType fromString(String type) {
    switch (type) {
      case 'bank':
        return AccountType.bank;
      case 'cash':
        return AccountType.cash;
      default:
        throw ArgumentError('Unknown AccountType: $type');
    }
  }

  String toJson() {
    switch (this) {
      case AccountType.bank:
        return 'bank';
      case AccountType.cash:
        return 'cash';
    }
  }

  String get displayName {
    switch (this) {
      case AccountType.bank:
        return 'Bank';
      case AccountType.cash:
        return 'Cash';
    }
  }
}

class Account {
  int? id;
  int? userId;
  AccountType? accountType;
  String? bankName;
  String? accountName;
  int? accountNumber;
  double? balance;
  double? credit;
  double? debit;

  Account({
    this.id,
    this.userId,
    this.accountType,
    this.bankName,
    this.accountName,
    this.accountNumber,
    this.credit,
    this.debit,
    this.balance,
  });

  factory Account.fromJson(Map<String, dynamic> data) => Account(
    id: data["id"],
    userId: data["user_id"],
    accountType: AccountTypeExtension.fromString(data["account_type"]),
    bankName: data["bank_name"] ?? "",
    accountName: data["account_name"] ?? "",
    accountNumber: data["account_number"] ?? 0,
    credit: (data["credit"] ?? 0).toDouble(),
    debit: (data["debit"] ?? 0).toDouble(),
    balance: (data["balance"] ?? 0).toDouble(),
  );

  Map<String, dynamic> toJson() => {
    "id": id,
    "user_id": userId,
    "account_type": accountType?.toJson(),
    "bank_name": bankName,
    "account_name": accountName,
    "account_number": accountNumber,
    "credit": credit,
    "debit": debit,
    "balance": balance,
  };
}
